### ISSUE TITLE
Question: How is inheritance working in RepoDB?

### ISSUE THREAD

**Issue by mediasystems-ch** [mikependon/RepoDb#955]
After studying the documentation, I have the impression that RepoDB supports inheritance. But I am not sure how it works exactly. For example, take the following example:

```C#
public class Vehicle { }

public class Car : Vehicle 
{
    public string Color { get; set; }
}
```

```C#
vehicle = new Car() { Color = "blue" };
var fields = FieldCache.Get(vehicle.GetType()).ToList();
var id = sqlConnection.Insert<Vehicle, int>(entity: vehicle, fields: fields);
```

Running this example results in the following exception:

**System.Data.SqlClient.SqlException: 'Must declare the scalar variable "@Color".'**

Apparently, the generic insert method only maps the properties from the base class.

What have I missed? How does querying work in the case of inheritance? How can I specify a discriminator column? Are there any examples?

Any help would be greatly appreciated.

**Comment by mikependon** [comment]
Error has happended that is because you passed the Vehicle into the TEntity explicitly. It should be a Car instead. The inference has removed the Color property since it is on the derived class.

**Comment by mediasystems-ch** [comment]
Yes, I suspected the same. Unfortunately, it is not possible for me to use the entity Car instead of Vehicle in my code. This would then invalidate all the advantages I gain from inheritance. Because then I would have to create a separate repository class for each derived entity class and reading Entities of type Vehicle would not work then either.

So this means that RepoDB doesn't support inheritance on entities, just like Entity Framework does?

**Comment by mikependon** [comment]
It does support it, but with your approach, it forces the type inference to the base class, which has eliminated the properties from the derived class.

Is it possible for you to just use the one below? (It will be inferred properly)

```csharp
vehicle = new Car() { Color = "blue" };
var id = (int)sqlConnection.Insert(vehicle);
```

Or, can you create a generic approach like below?

```csharp
public TResult Insert<TEntity, TResult>(TEntity entity) where TEntity : Vehicle =>
      connection.Insert<TEntity, TResult>(entity);
```

And call it like below?

```csharp
vehicle = new Car() { Color = "blue" };
var id = Insert(vehicle);
```

--------------

Though with your approach above, it is considered a separate scenario as it does inspect the type properties based on the passed-instance and not from the T generic type. (This is least scenario in use, but I will be frank that I had used this when inferring the Dictionary and ExpandoObject).

**Comment by mediasystems-ch** [comment]
Thanks for your answer. If I use this ...

```C#
var id = (int)sqlConnection.Insert(entity: vehicle, fields: fields);
```
... then I'm getting the same exception:

**System.Data.SqlClient.SqlException: 'Must declare the scalar variable "@color".'**

Which is also not clear to me how this works when reading. How can RepoDB distinguish which sub-entity it is? You should be able to specify a discriminator column in some way. Is this possible?

**Comment by mikependon** [comment]
Really, does the code below failing? _(Without the `fields` argument for this case)_

```csharp
var vehicle = new Car() { Color = "blue" };
var id = (int)sqlConnection.Insert(vehicle);
```

But, it should still work even with the `fields`.

**Comment by mikependon** [comment]
This one I had never tested really. Is this your scenario? (But it seems this does not really inferring the proper Car type)

```csharp
Vehicle vehicle = new Car() { Color = "blue" };
var id = (int)sqlConnection.Insert(vehicle);
```

**Comment by mediasystems-ch** [comment]
Yes, that's my scenario. I'm not able to call the Insert method without the fields argument because auf the optimistic concurrency implementation. This would lead then into a different exception.

**Comment by mikependon** [comment]
I guess it would work if you declare the `car` instance like below.

```csharp
var vehicle = new Car() { Color = "blue" };
// Or
// Car vehicle = new Car() { Color = "blue" };
```

Then, the type inference for the base class is seems a bug to me.

**Comment by mediasystems-ch** [comment]
Using
```c#
Car vehicle = new Car() { Color = "blue" };
```
is not possible in my scenario.

**Comment by mikependon** [comment]
Just an update to this. RepoDB does the generic type inference via `typeof(T)`. The type inferrence is very dependent to the type of passed 'variable' on this regards. See below.

Let us say you have this method.

```csharp
private string GetTypeName<T>(car) =>
     typeof(T).Name;
```

The code below simulates this scenario.

```csharp
var car = new Car();
var name = GetTypeName(car); // Returns 'Car'

Vehicle car = new Car();
var name = GetTypeName(car); // Returns 'Vehicle'
```

The culprit is on the `typeof(T)` class. We will do the change by calling the `instance.GetType()` prior to the generic type inference.

Note: The impact pertaining to this change is quite big as we need to scan the entire solution with such scenario.

**Comment by mikependon** [comment]
@mediasystems-ch - we have started looking at this last night and it is true that the impact is huge because of the approach we wrote when extracting the generic types. This affect the major operations of the library, primarily the Insert, InsertAll, Merge, MergeAll, Update and UpdateAll. The required code changes are huge pertaining to this, therefore, it may take a little while for us to deliver this. But, this will be a part of the next release, probably in the next 2 weeks. All the impacted use-cases are all covered by the Integration Tests.
